libbeat/common/transport/logging.go

```diff
@@ -0,0 +1,52 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package transport
+
+import (
+	"io"
+	"net"
+
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+type loggingConn struct {
+	net.Conn
+	logger *logp.Logger
+}
+
+func LoggingDialer(d Dialer, logger *logp.Logger) Dialer {
+	return ConnWrapper(d, func(c net.Conn) net.Conn {
```

Comment:
One more ask: let's make use of structured logging, so we can tell which host the log output is to be associated with. e.g.

```
return DialFunc(func(network, address string) (net.Conn, error) {
  logger := logger.With("network", network, "address", address)
  conn, err := d.Dial(network, address)
  ...
})
```

The address includes `<hostname|ip>:<port>`.